### PR TITLE
Channels

### PR DESCRIPTION
**Pull request by w0de**
First pass at implementing "channels" to allow variation in promotion schedule. Looking for feedback @ChefAustin - a solution to #25

### CODE REVIEW

**Comment by ChefAustin** [comment]
Looping in @erikng for his thoughts, too.

**Comment by erikng** [comment]
The main function returns an int and a float. Perhaps you should round the float to an integer so the type is static?

**Comment by w0de** [comment]
> The main function returns an int and a float. Perhaps you should round the float to an integer so the type is static?

The function always returns a float. It is important that is is a float, since this implementation expects <1.0 for _faster_ channels. (The promotion period is a function of the channel multiplier, a multiplier of 0.5 will result in a half-length promotion period.)

**Comment by erikng** [comment]
The function has a `return 1`

File: autopromote/README.md [hunk 1]
```diff
@@ -41,3 +41,9 @@ Format: `{"hour": int, "minute": int}`
 __enforce_force_install_time__: Have you decided 4:30 is a bad force install time? Set this value and all pkginfos, once parsed, will have their force_install_after_date changed to reflect the it.
 
 __force_install_denylist__: A list of pkginfos (as defined in their `name` attribute) on which no force_install_after_date will ever be set.
+
+__channels__: Channels allow one to specify a faster or slower promotion schedule for specific packages. This is a dictionary of channel names and an int or float multiplier:
+
+`{"channels": {"slow": 2.5}}` - this channel configuration would cause any packages in the "slow" channel to be promoted 2.5 times *slower*. This is achieved by multiplying the current promotion period by the channel's value. For faster promotion schedules, specify a float modifier of less than 1.
```
Comment (ChefAustin) [inline autopromote/README.md:47]: Just for my own clarity: This means that a channel configuration like this...

```
{"channels": {"fast": 0.5}}
```
...would make any items associated to that channel to be promoted at a rate that is 2x _faster_ than the base-level promotion schedule?

If that is, indeed, the case, maybe you could add a quick little ditty like that ☝️ to this `README.md` to explicitly denote this type of configuration for hastening a promotion cadence?

File: autopromote/autopromote.py [hunk 2]
```diff
@@ -211,6 +211,20 @@ def get_next_catalog(latest_catalog):
     return None
 
 
+def get_channel_multiplier(plist):
+    """Retrieve the float multiplier for plist's channel. Returns multiplier or 1"""
+
+    channel = plist.get("_metadata", {}).get("channel")
+    if channel is None:
+        return 1
```
Comment (ChefAustin) [inline autopromote/autopromote.py:219]: I think this line (and L223, perhaps) is what @erikng was referring to with [his comment on this PR](https://github.com/Gusto/it-cpe-opensource/pull/26#issuecomment-652065962).

Should this be `return 1.0` for consistent types?

**Comment by w0de** [comment]
> The function has a `return 1`

Haha, good point. I'll make that a float.

File: autopromote/README.md (hunk 1 above)
Comment (w0de) [reply autopromote/README.md]: Correct - added a note clarifying this.

**Pull request closed without merging**